samples/HttpServer_FirmwareUpload/app/SignedRbootOutputStream.cpp

```diff
@@ -35,23 +34,21 @@ size_t SignedRbootOutputStream::write(const uint8_t* data, size_t size)
 				if(unexpectedMagic || unexpectedLoadAddress) {
 					setError(unexpectedLoadAddress ? "Unexpected load address. Try image for other slot."
 												   : "Invalid image received.");
-					return 0;
+				} else {
+					init();
 				}
-
-				init();
 			}
 		}
 
-		if(size > 0) {
-			crypto_sign_update(&verifierState, static_cast<const unsigned char*>(data), size);
-			const size_t written = RbootOutputStream::write(data, size);
-			consumed += written;
-			if(written != size) {
+		if(!errorFlag && (available > 0)) {
+			crypto_sign_update(&verifierState, static_cast<const unsigned char*>(data), available);
+			const size_t written = RbootOutputStream::write(data, available);
+			if(written != available) {
 				setError("Flash write failure");
 			}
 		}
 	}
-	return consumed;
+	return size; // Always pretend to consume everything. Otherwise the connection is closed before we get a chance to report our error message.
```

Comment:
> Otherwise the connection is closed before we get a chance to report our error message.

Not really. The onMessageComplete callback will be called. In it you can decide what to do. If you return non-zero value the default value error page will be returned but if you return zero you can send custom error response.

For example the code below:
```
int onUpload(HttpServerConnection& connection, HttpRequest& request, HttpResponse& response)
{
	if(!response.isSuccess()) {
		debug_e("Request failed.");
		return -1;
	}

	ReadWriteStream* file = request.files["firmware"];
	if(file == nullptr) {
		debug_e("Something went wrong with the file upload");
		return 1;
	}

	SignedRbootOutputStream* uploadStream = static_cast<SignedRbootOutputStream*>(file);

	if(uploadStream->verifySignature(firmwareVerificationKey)) {
		// ...
	} else {
		response.code = HTTP_STATUS_BAD_REQUEST;
		response.setContentType(MIME_HTML);
		String html = "<H2 color='#444'>" + uploadStream->errorMessage + "</H2>";
		response.headers[HTTP_HEADER_CONTENT_LENGTH] = html.length();
		response.sendString(html);
	}
	return 0;
}
```

Can be changed to:

```
int onUpload(HttpServerConnection& connection, HttpRequest& request, HttpResponse& response)
{
	bool success = response.isSuccess();
        String errorMessage = "Bad Request";

        if(success) {
            ReadWriteStream* file = request.files["firmware"];
    	    if(file == nullptr) {
		debug_e("Something went wrong with the file upload");
                errorMessage = "<.....>";
		success = false;
	     }
        }

	if(success) {
             SignedRbootOutputStream* uploadStream = static_cast<SignedRbootOutputStream*>(file);
             if(uploadStream->verifySignature(firmwareVerificationKey)) {
		//...
	     }
             else {
                 success = false;
                 errorMessage =  uploadStream->errorMessage;
             }
        }
	
       if(!success) {
		response.code = HTTP_STATUS_BAD_REQUEST;
		response.setContentType(MIME_HTML);
		String html = "<H2 color='#444'>" + errorMessage + "</H2>";
		response.headers[HTTP_HEADER_CONTENT_LENGTH] = html.length();
		response.sendString(html);
       }

       return 0;
}
```

Reply:
Thanks for the explanation, but unfortunately, I cannot to get it to work your way either.
Digging deeper through the HTTP server code, the reason seems to be that `HttpConnection::onTcpReceive` returns `false` in case of an HTTP parser error (and `SignedRbootOutputStream::write` returning something different from its `size` argument *is* considered a parser error):
```	
	int parsedBytes = http_parser_execute(&parser, &parserSettings, data, size);
	if(HTTP_PARSER_ERRNO(&parser) != HPE_OK) {
		// we ran into trouble - abort the connection
		onHttpError(HTTP_PARSER_ERRNO(&parser));
		return false;
	}
```
This causes the TCP connection to close, even if the `onMessageComplete` callback invoked from `onHttpError` has schedules an appropriate error response.

Reply:
> This causes the TCP connection to close, even if the onMessageComplete callback ...

That should be taken care of in PR #1820.

Reply:
Yeah that works. I'll rebase when yor PR is merged.

Reply:
> Yeah that works. I'll rebase when yor PR is merged.

Great. The PR is merged so you can rebase.

Reply:
Arrgh, my confirmation came too early (had the workaround still in place).

It does not work, because (note numbered comments)...

```
bool HttpConnection::onTcpReceive(TcpClient& client, char* data, int size)
{
	if(HTTP_PARSER_ERRNO(&parser) != HPE_OK) {
		setTimeOut(1);
		// if the parser is in error state then just ignore the incoming data.
		return false; // (3) kicks in for chunks received *after* graceful error handling
	}

	int parsedBytes = http_parser_execute(&parser, &parserSettings, data, size);
	if(HTTP_PARSER_ERRNO(&parser) != HPE_OK) {
		bool isRecoverable = onHttpError(HTTP_PARSER_ERRNO(&parser));
		if(!isRecoverable) {
			// we ran into trouble - abort the connection
			setTimeOut(1);
			return false;
		}
		// return true; <-- (1) Need this, or...
	}

	if(parser.upgrade) {
		return onProtocolUpgrade(&parser);
	} else if(parsedBytes != size) {
		return false; // (2) ...this kicks in
	}

	return true;
}
``` 
(1) missing `return true` in this line. Otherwise, method continues and...
(2) ...the partially consumed chunk is detected in this branch and handled ungracefully.

But even if these points are addressed and the connection stays open, any subsequently received chunk of data steps into.. 

(3) this condition, where the connection is again closed immediately.

Reply:
@aemseemann Can you rebase this PR on the latest changes in `develop` and try again? It should work now as expected.

Reply:
@slaff: Thanks for improving the error handling. Unfortunately, the connection is still closed immediately when uploading the "wrong" file. I'll further investigate it tomorrow.